### ISSUE TITLE
IRaftEventListener.LeaderUpdated Called Frequently

### ISSUE THREAD

**Issue by jkassis** [lni/dragonboat#139]
My cluster seems bootstrapped into three healthy nodes, but i get very frequent updates on this channel.

![image](https://user-images.githubusercontent.com/3630156/91399857-94ef4780-e7f3-11ea-820c-143246398aaf.png)

It seems that the leader does not get notifications when it is still the leader, but followers get frequent notifications that they are not the leader. Once per heartbeat?

Normal?

**Comment by jkassis** [comment]
I'm now have dragonboat integration to the point that i can actually run my services... that's amazing. I swapped out hashicorp/raft for dragonboat in one day... plus maybe 4-5 hours of research. thanks for great apis, docs, and support.

not sure if it's related to the chatty LeaderUpdates, but the performance of my dragonboat raft integration is about 50% of the performance of my hashicorp/raft integration right now. my db initializes in 4secs vs 2secs.

some important differences i can think of...

hashicorp/raft:
badger raft persistence
different log adapter

dragonboat:
![image](https://user-images.githubusercontent.com/3630156/91409396-34173d80-e7fa-11ea-96ac-ee33a62654f5.png)

**Comment by lni** [comment]
@jkassis LeaderUpdated() is indeed called more frequent than it should. Thanks for reporting this issue and it will be addressed soon. 

For the performance issue you mentioned above, my understanding is you observed the startup time using Dragonboat is 200% when compared to your old implementation using hasicorp Raft and badger?

**Comment by jkassis** [comment]
Well startup time is perhaps another issue. I see sometimes long ~30s startup times with an empty database which is an issue for developer time / inner loop, but this is regarding my service.

It's a db that i initialize (schema and initial objects) through an external command. The external command runs notably slower... ~2x and is clearly reproducible by swapping out hashi / dragonboat.

Here is my Proposal code which is currently calling the Async version of Propose. Does it provide any benefit since I'm immediately blocking on the CompletedC?

![image](https://user-images.githubusercontent.com/3630156/91462825-7adb5680-e83f-11ea-850a-4490a1a736a1.png)

**Comment by jkassis** [comment]
Here's a little more info, though perhaps this belongs in a separate issue at this point:

```
INFO[0082] Aug 27 11:52:43.743925000 rrie/core/kittie/serviceRaftDB.go:  31  go version: go1.14                           
INFO[0082] Aug 27 11:52:43.743948000 rrie/core/kittie/serviceRaftDB.go:  31  dragonboat version: 3.2.6 (Rel)
INFO[0082] Aug 27 11:52:43.743966000 rrie/core/kittie/serviceRaftDB.go:  31  raft entry encoding scheme: colfer           
WARN[0082] Aug 27 11:52:43.743992000 rrie/core/kittie/serviceRaftDB.go:  32  running on darwin, don't use for production purposes 
INFO[0082] Aug 27 11:52:43.744050000 rrie/core/kittie/serviceRaftDB.go:  31  system ticker precision is set to 1ms (default) 
WARN[0082] Aug 27 11:52:43.747876000 rrie/core/kittie/serviceRaftDB.go:  32  DeploymentID not set in NodeHostConfig, default to 1 
INFO[0082] Aug 27 11:52:43.747905000 rrie/core/kittie/serviceRaftDB.go:  31  DeploymentID set to 1                        
INFO[0082] Aug 27 11:52:43.800543000 rrie/core/kittie/serviceRaftDB.go:  31  using plain ShardedRDB                       
WARN[0082] Aug 27 11:52:43.800597000 rrie/core/kittie/serviceRaftDB.go:  32  pebble support is experimental, DO NOT USE IN PRODUCTION 
INFO[0083] Aug 27 11:52:44.897414000 rrie/core/kittie/serviceRaftDB.go:  31  using plain ShardedRDB                       
INFO[0084] Aug 27 11:52:46.607602000 rrie/core/kittie/serviceRaftDB.go:  31  transport type: go-tcp-transport             
INFO[0084] Aug 27 11:52:46.607629000 rrie/core/kittie/serviceRaftDB.go:  31  logdb type: sharded-pebble                   
INFO[0084] Aug 27 11:52:46.607646000 rrie/core/kittie/serviceRaftDB.go:  31  nodehost address: 127.0.0.1:9999             
WARN[0082] Aug 27 11:52:43.743844000 rrie/core/kittie/serviceRaftDB.go: 130  ServiceRaft: Play: Starting wht NodeHostConfig: {"EnableMetrics":true,"ListenAddress":"localhost:9999","NodeHostDir":"/var/multi/cluster/Jeremys-MBP.hsd1.ca.comcast.net-server-0/raft/dbJeremys-MBP.hsd1.ca.comcast.net-server-0","RTTMillisecond":200,"RaftAddress":"127.0.0.1:9999","WALDir":"/var/multi/cluster/Jeremys-MBP.hsd1.ca.comcast.net-server-0/raft/dbJeremys-MBP.hsd1.ca.comcast.net-server-0"} 
WARN[0084] Aug 27 11:52:46.607777000 rrie/core/kittie/serviceRaftDB.go: 152  ServiceRaft: Play: Starting with NodeConfig: {"NodeID":10492612360936044872,"ClusterID":10,"CheckQuorum":true,"ElectionRTT":10,"HeartbeatRTT":1,"SnapshotEntries":2048,"CompactionOverhead":10,"OrderedConfigChange":false,"MaxInMemLogSize":0,"SnapshotCompressionType":0,"EntryCompressionType":0,"DisableAutoCompactions":false,"IsObserver":false,"IsWitness":false,"Quiesce":false} 
```

**Comment by jkassis** [comment]
Does Hashi RAFT return on RAFT WAL commit or settling of the TXN?

**Comment by jkassis** [comment]
today i updated my code so that i can switch back and forth between hashicorp and dragonboat with a single boolean. dragonboat seems to perform 10x worse in my 3 NodeHost 1 Cluster (single raft) test case. the one thing that isn't real-world about *this* test is that all of the NodeHosts are run in a single process, just bound to different ports. To be fair this is true for Hashicorp as well. In any case... here is a snapshot of my testresults. I've done my best to isolate the variables of course, with the boolean flipping. I'll experiment further, but welcome any thoughts. An order of magnitude seems like a misconfiguration.

![image](https://user-images.githubusercontent.com/3630156/91508582-c49e5e00-e88c-11ea-88c4-90d53b058a7f.png)

**Comment by jkassis** [comment]
regarding the long startup times... after further investigation this is *not* a dragonboat issue. it's clearly on myside where i'm initing the DB for my rafted service. the performance i'm seeing here obviously *is* a concern though.

thanks again!

**Comment by lni** [comment]
@jkassis 

I am still waiting for your definition on "DB init" - when your screenshot shows that it takes the hashicorp implementation less than 1 second to complete the DB init step but a few seconds for dragonboat to do the same, what steps are involved in your DB init steps? Does that a few seconds spent by dragonboat includes its start up time? 

In the screenshot above, that dbInit script got executed multiple times, did that script recreate a new DB instance and a new instance (with its previous state all removed) of dragonboat every time? 

From the dragonboat log you posted, it costs a few seconds for the dragonboat to start up, is that is cause?

**Comment by jkassis** [comment]
hi @lni. i think the name dbInit is a little misleading in this case. it just does raft proposals. mostly writes. my rafted service is a custom db that uses a local badger store for kv persistence.

one thing that is a little different come to think of it is that the hashicorp raft is using the raft-badger adapter to manage the WAL in the same badger instance as my rafted service. doesn't seem like that should be a factor, but its an important architectural difference here.

the dbInit.js script is a node client that sends one HTTP rest request to a node HTTP server which has all my business logic. the server handles the dbInit request by making multiple HTTP requests to the rafted service. some to create schema objects in my custom DB and then to create initial objects. the schema operations don't hit the raft, but the object writes do of course.

unless you have some quick ideas, i would probably have to do some real benchmarking and profiling at this point.

i'm interested in dragonboat for its multi-raft capability to handle shards for my db which i anticipate needing soon.

**Comment by lni** [comment]
@jkassis you didn't answer my question - when you claim that a few proposals can take a few extra seconds to complete when compared with hashicorp raft, did those a few extra seconds include dragonboat's start up time? In particular, does your dbInit.js script start a brand new dragonboat instance each time? Based on the log you provided above, it took dragonboat around 3 seconds to start when your screenshot shows 3-5 seconds to complete that dbInit.js script. The obvious issue here is are you measuring start up time or are you measuring the time it took to complete a couple proposals? 

It would also be much easier if you can do some timing on your server side and provide the break down on how those few seconds were spent, e.g. X milliseconds on dragonboat.NewNodeHost(), Y milliseconds to wait for the leader to be elected followed by Z milliseconds on completing 3 proposals.

**Comment by jkassis** [comment]
@lni. sorry it wasn't clear. dragonboat is started once when the db server process comes up, so this is definitely. not startup time over and over. here's what I see in a stackdriver profile...

![image](https://user-images.githubusercontent.com/3630156/91593527-33bf9500-e915-11ea-86aa-0131d8e590f3.png)

the process spends lots of time in park which indicates blocked goroutines. MTU on osxdarwin is 1500. this is all going through loopback i would think. my write txns are *farily* big but under 1k i'm sure. pebble logging is a little more than i would expect.

i'll do a snapshot for HC for comparison.

**Comment by lni** [comment]
A few issues here - 

1. given your state machine states are persisted to disk, you should probably use the OnDiskStateMachine type. See the example project on how to use it. 
2. you probably don't need to take snapshot very often as your state is always on disk. Normally only do that when you need to compact raft logs. what kind of performance you can get if snapshotting is completely disabled (e.g. use a very large snapshot threshold)? 
3. the Pebble logging part you highlighted on the above profiling result is Pebble's WAL writer, Pebble's WAL write is very efficient, if it is taking much more resources than the one used in other implementations - be careful, the other one might be cutting corns. 

The log you posted above also indicates that you probably be using a mbp, mbp has extremely slow fsync(), it is a terrible platform for measuring any database/consensus library related performance. To give you some quick numbers - a subset of dragonboat's tests (make test-multiraft) can complete on a 5 years old server with a 6 years old SSD in 20-30 seconds, but it takes >300 seconds on my mbp.

**Comment by lni** [comment]
It will also be useful to know how many proposals will be triggered each time when that dbInit.js script is invoked?

**Comment by jkassis** [comment]
snapshotting is run every 2048 logs so i was able to get profiles that exclude snapshotting. here they are head to head...

The DragonBoat profile on top is averaged over 7 profiles. The HashiCorp profile on bottom is averaged over 11 profiles.

Total CPU used by DB is less than HC while TPS is lower. this indicates lots of blocking i think. My statemachine is not *entirely* ondisk, which is why i didn't go there initially, but i'll look into it more closely. i'm guessing that would reduce blocking.

![image](https://user-images.githubusercontent.com/3630156/91600308-9a947c80-e91c-11ea-9ee8-5ba98ba20e7a.png)

\# of proposals per dbInit is 34

**Comment by jkassis** [comment]
the select statements in nodeWorkerMain and taskWorkerMain are the hotspots AFAICT.

![image](https://user-images.githubusercontent.com/3630156/91623411-460af480-e950-11ea-97c4-e1d6b692aa5a.png)

**Comment by jkassis** [comment]
hashicorp does not contribute to block time like this...

![image](https://user-images.githubusercontent.com/3630156/91623699-48218300-e951-11ea-98b0-0282796726c6.png)

**Comment by jkassis** [comment]
agree mbp is not the best, but for now it's apples to apples since both benchmarks are running on the same machine.

**Comment by lni** [comment]
@jkassis those block time info are not very useful, nodeWorkerMain and taskWorkerMain are blocked on select statements when they are idle. 

Do you have everything including your server side code running on your mbp? If that is the case, please add some logging to see when (1) HTTP request is received, (2) proposal is made, (3) IStateMachine.Update() is called, (4) IStateMachine.Update() is ready to return, (5) proposal is completed with Update()'s returned result passed to the client, (6) HTTP response is sent and received.

When everything is running on a single mbp, you should be able to obtain the above logs with timestamps, then we can see how time is spent.

**Comment by lni** [comment]
Dragonboat actually has a built-in tool to check the problem you reported. It is available at - 

https://github.com/lni/dragonboat/tree/master/tools/checkdisk

when I run the master HEAD version on my mbp 2018 using the following command, it will create two NodeHost instances with 48 raft clusters. A single client is created to keep making proposals on one of those 48 clusters, it keeps doing that for 60 seconds and report how many proposals per second it completed.

```
./checkdisk -num-of-clients 1 -two-nodehost
```
I got roughly 30 millisecond per proposal. 
```
total 1920, 32 proposals per second
```

**That is already several times faster than the Dragonboat numbers you reported.**  

On a 4-5 years old server with a proper SSD, the same tool using the same command got me the following result 

```
clusters are ready, will run for 60 seconds
read 0, 0 reads per second
total 109920, 1832 proposals per second
``` 

That is about 0.5ms per proposal. 

You may still argue that such single raft "performance" is still slightly lower than other single raft implementation, then please feel free to use 128 or 1024 raft clusters and compare the numbers. ;)

To solve the high latency issue you experienced, I think you should build a minimum program that can reproduce the slow proposal issue and share the code of that so I can actually look into the reported problem. Based on the above described tools and results, I don't see any issue here - the slowness you experienced might be caused by your system, your own code or another 3rd party library.

**Comment by jkassis** [comment]
trying...
`
[I] jkassis@Jeremys-MBP ~/c/dragonboat (release-3.2) [2]> go build github.com/lni/dragonboat/v3/tools/checkdisk                                                                                                                                         08.29 12:37
# github.com/lni/dragonboat/v3/internal/logdb/kv/rocksdb/gorocksdb
internal/logdb/kv/rocksdb/gorocksdb/backup.go:90:8: could not determine kind of name for C.free
internal/logdb/kv/rocksdb/gorocksdb/backup.go:152:2: could not determine kind of name for C.rocksdb_backup_engine_close
internal/logdb/kv/rocksdb/gorocksdb/backup.go:113:2: could not determine kind of name for C.rocksdb_backup_engine_create_new_backup
internal/logdb/kv/rocksdb/gorocksdb/backup.go:126:6: could not determine kind of name for C.rocksdb_backup_engine_get_backup_info
internal/logdb/kv/rocksdb/gorocksdb/backup.go:33:15: could not determine kind of name for C.rocksdb_backup_engine_info_backup_id
internal/logdb/kv/rocksdb/gorocksdb/backup.go:22:13: could not determine kind of name for C.rocksdb_backup_engine_info_count
internal/logdb/kv/rocksdb/gorocksdb/backup.go:48:2: could not determine kind of name for C.rocksdb_backup_engine_info_destroy
internal/logdb/kv/rocksdb/gorocksdb/backup.go:43:15: could not determine kind of name for C.rocksdb_backup_engine_info_number_files
internal/logdb/kv/rocksdb/gorocksdb/backup.go:38:15: could not determine kind of name for C.rocksdb_backup_engine_info_size
internal/logdb/kv/rocksdb/gorocksdb/backup.go:17:5: could not determine kind of name for C.rocksdb_backup_engine_info_t
internal/logdb/kv/rocksdb/gorocksdb/backup.go:27:15: could not determine kind of name for C.rocksdb_backup_engine_info_timestamp
internal/logdb/kv/rocksdb/gorocksdb/backup.go:92:8: could not determine kind of name for C.rocksdb_backup_engine_open
internal/logdb/kv/rocksdb/gorocksdb/backup.go:141:2: could not determine kind of name for C.rocksdb_backup_engine_restore_db_from_latest_backup
internal/logdb/kv/rocksdb/gorocksdb/backup.go:81:8: could not determine kind of name for C.rocksdb_backup_engine_t
internal/logdb/kv/rocksdb/gorocksdb/backup.go:61:6: could not determine kind of name for C.rocksdb_restore_options_create
internal/logdb/kv/rocksdb/gorocksdb/backup.go:75:2: could not determine kind of name for C.rocksdb_restore_options_destroy
internal/logdb/kv/rocksdb/gorocksdb/backup.go:70:2: could not determine kind of name for C.rocksdb_restore_options_set_keep_log_files
internal/logdb/kv/rocksdb/gorocksdb/backup.go:55:5: could not determine kind of name for C.rocksdb_restore_options_t
`

**Comment by jkassis** [comment]
works on master... first run...

```
clusters are ready, will run for 60 seconds
read 0, 0 reads per second
total 1888, 31 proposals per second
2020-08-29 12:54:04.246265 W | transport: localhost:26000 became unreachable, affecting 48 raft nodes
2020-08-29 12:54:04.246301 W | transport: localhost:26000 became unreachable, affecting 48 raft nodes
2020-08-29 12:54:04.246375 W | transport: localhost:26000 became unreachable, affecting 48 raft nodes
2020-08-29 12:54:04.246494 W | transport: localhost:26000 became unreachable, affecting 48 raft nodes
2020-08-29 12:54:04.252890 W | transport: localhost:26001 became unreachable, affecting 48 raft nodes
2020-08-29 12:54:04.252907 W | transport: localhost:26001 became unreachable, affecting 48 raft nodes
2020-08-29 12:54:04.252925 W | transport: localhost:26001 became unreachable, affecting 48 raft nodes
2020-08-29 12:54:04.252937 W | transport: localhost:26001 became unreachable, affecting 48 raft nodes
```

**Comment by jkassis** [comment]
not sure if the broken pipe is normal but 41 proposals per second with a total of 2496 looks good...
```
clusters are ready, will run for 60 seconds
read 0, 0 reads per second
total 2496, 41 proposals per second
2020-08-29 12:56:15.396411 W | transport: localhost:26000 became unreachable, affecting 48 raft nodes
2020-08-29 12:56:15.396480 W | transport: localhost:26000 became unreachable, affecting 48 raft nodes
2020-08-29 12:56:15.396590 W | transport: localhost:26000 became unreachable, affecting 48 raft nodes
2020-08-29 12:56:15.396639 W | transport: localhost:26000 became unreachable, affecting 48 raft nodes
2020-08-29 12:56:15.401309 E | transport: send batch failed, target [00009:00002] (write tcp 127.0.0.1:54202->127.0.0.1:26001: write: broken pipe), 1
2020-08-29 12:56:15.401325 W | transport: breaker localhost:26000 to localhost:26001 failed, connect and process failed: write tcp 127.0.0.1:54202->127.0.0.1:26001: write: broken pipe
2020-08-29 12:56:15.401337 W | transport: localhost:26001 became unreachable, affecting 48 raft nodes
2020-08-29 12:56:15.416640 W | transport: localhost:26001 became unreachable, affecting 48 raft nodes
2020-08-29 12:56:15.416654 W | transport: localhost:26001 became unreachable, affecting 48 raft nodes
2020-08-29 12:56:15.416662 W | transport: localhost:26001 became unreachable, affecting 48 raft nodes
```

**Comment by jkassis** [comment]
added prometheus histograms to http time, raft proposal time, and service time. eliminated service time with a noop.

![image](https://user-images.githubusercontent.com/3630156/91668567-74a8dc80-eac2-11ea-9d4c-a7e483ebd6fc.png)

**Comment by jkassis** [comment]
![image](https://user-images.githubusercontent.com/3630156/91668792-41ffe380-eac4-11ea-980b-eb28a50b53db.png)

**Comment by jkassis** [comment]
i tried a few things... cut TTL in half,  disable metrics, set quorum checking == false. no leverage from any of these settings. 65ms * 34 proposals == 2.2secs which is about what i'm seeing. the stats from checkdisk are a little better, but still not as fast as hashicorp. the only 2 things i can think of is that i'm using raft-badger for the WAL on the hashicorp while db uses pebble. either that or there is some blocking / queueing code in db that is better for multi-raft but not so good for single raft. 

![image](https://user-images.githubusercontent.com/3630156/91669089-0b779800-eac7-11ea-8655-18a510a4939a.png)

**Comment by jkassis** [comment]
here's my badger config for the hashicorp WAL db
![image](https://user-images.githubusercontent.com/3630156/91669151-abcdbc80-eac7-11ea-90ff-2399375f6421.png)

**Comment by jkassis** [comment]
i don't see dragonboat metrics in prometheus. a similar histogram around WAL DB commit time would be helpful.

**Comment by lni** [comment]
@jkassis 

Let's be crystal clear first, there are two issues so far - 

1. Your closed source program with a ton of other features, business logics, dependencies and communication overheads (e.g. HTTP) are costing 4.58 seconds on average for 34 proposals, that is 135ms per proposal. I've provided you a simple program clearly demonstrating that a proposal can be completed on your machine in about 34ms. How your closed source system with extra business logics and dependencies took much longer to handle a request is up to you for further investigation. As explained, it might be caused by bugs in your code, in some 3rd party libraries or maybe even by your own measurement mistakes. 

If you still believe such 100ms extra delay is caused by Dragonboat, **please provide a minimum program that can actually reproduce your claimed delay.** 

2. That 34ms delay reported by checkdisk program is indeed still higher than what you get from hashicorp in a **single client single raft** scenario because Dragonboat intentionally slow down certain response to allow more client requests to be batched together for efficient concurrent accesses. 

I have a working feature branch with minimum changes (around 10 LOC) that can reduce such delay in a **single client single raft** scenario by half, making its delay the exact same as the hashicorp Raft. However, a bold however, that is not a fix or optimization, it removes certain batching considered useful to just score better in a meaningless benchmark. In my tests, I saw such change reducing the delay to half when there is **one client and one Raft group**, but throughput doesn't scale well when there are dozens to hundreds concurrent clients. 

See the log below, using the above mentioned changes, a single client on a single Raft can complete 67 proposals per second (15ms delay) on mbp's extremely slow SSD. This is half of the delay you saw in your own test on a mbp. 

```
lni$ ./checkdisk -num-of-clients 1 -num-of-clusters 1 -two-nodehosts
2020-08-31 14:00:57.176405 W | dragonboat: running on darwin, don't use for production purposes
2020-08-31 14:00:57.176585 I | config: system ticker precision is set to 1ms (default)
2020-08-31 14:00:57.176602 I | config: using default ExpertConfig
2020-08-31 14:00:57.212834 W | pebblekv: pebble support is experimental, DO NOT USE IN PRODUCTION
2020-08-31 14:01:01.448790 W | dragonboat: running on darwin, don't use for production purposes
2020-08-31 14:01:01.448868 I | config: system ticker precision is set to 1ms (default)
2020-08-31 14:01:01.448880 I | config: using default ExpertConfig
2020-08-31 14:01:07.649187 W | raft: [f:1,l:2,t:1,c:2,a:2] [00001:00001] t2 became candidate
2020-08-31 14:01:07.649221 W | raft: [f:1,l:2,t:1,c:2,a:2] [00001:00001] t2 received RequestVoteResp from n00001
2020-08-31 14:01:07.649240 W | raft: [f:1,l:2,t:1,c:2,a:2] [00001:00001] t2 sent RequestVote to n00002
2020-08-31 14:01:07.661875 W | raft: [f:1,l:2,t:1,c:2,a:2] [00001:00002] t1 received RequestVote with higher term (2) from n00001
2020-08-31 14:01:07.661894 W | raft: [f:1,l:2,t:1,c:2,a:2] [00001:00002] t1 become followerKE after receiving higher term from n00001
2020-08-31 14:01:07.661907 W | raft: [f:1,l:2,t:1,c:2,a:2] [00001:00002] t2 cast vote from n00001 index 2 term 2, log term: 1
2020-08-31 14:01:07.673450 W | raft: [f:1,l:2,t:1,c:2,a:2] [00001:00001] t2 received RequestVoteResp from n00002
2020-08-31 14:01:07.673471 W | raft: [f:1,l:2,t:1,c:2,a:2] [00001:00001] t2 received 2 votes and 0 rejections, quorum is 2
clusters are ready, will run for 60 seconds
read 0, 0 reads per second
total 4064, 67 proposals per second
```

You have been trying to measure dragonboat's "performance" in a very strange way. You first ignored the most significant contributor of the delay which is the 100ms extra cost described in (1) caused by factors outside dragonboat. Your measurements then focused on **one client on one Raft group** when all use cases I've seen/heard rely on multi Raft or at least multi clients performance. To me it is like you are focusing on the fact that a Boeing aircraft can't move as fast as a car on the runway.

**Comment by jkassis** [comment]
@lni first let me thank you again for your responsiveness, thoroughness, and support. i still do believe that dragonboat is an excellent product and that you are personally an exceptional engineer. What you have done with dragonboat is very special and i honor your work.

I do believe i understand your analysis and perspective and i think we are getting to the bottom of the mystery. Part of the challenge for me has been to understand what use-cases dragonboat handles the best.

My goal has been to asses whether I can use Dragonboat as my RAFT solution from the beginning and as my architecture evolves to use multi-raft. I would rather use and gain experience with Dragonboat than introduce it later during live operation of my service, or worse... build my own multi-raft solution.

As i look at the data i've gathered, it seems to support the advertised performance of dragonboat... here is the chart currently provided on the homepage...

![image](https://user-images.githubusercontent.com/3630156/91754512-615b4700-eb7e-11ea-9f6f-88939ae21a5b.png)

The two results from the checkdisk tool posted above present average latency of 32ms (31 proposals per sec) and 24ms (41 proposals per sec) which is mostly consistent with that chart.

FWIW I know that my 2019 MacBook Pro does not have the fastest SSD on the market, but I am personally not concerned with the raw numbers, just the comparison of latencies between RAFT implementations on equivalent hardware.

The results I posted from my application tell a mostly consistent story I believe. It shows raft latency that is ~ 2x greater than the checkdisk results, but that is perhaps attributable to the single-raft / single-client use case i'm testing now. Perfectly reasonable.

Please note that I eliminated calls to my service as evidenced in the Prometheus metrics posted above. The RAFT proposals are sent to the nodes (there are 3x as many service calls as there are proposals), but the service performs a NOOP. The histogram data on the right of both of those graphics show that nearly all calls to the NOOP service complete in less than 1000ns, contributing negligible time.

Also note from the Prometheus metrics that the HTTP service handler (the front-end for the RAFT proposals) takes negligible time to complete, so this is a measure of RAFT proposal processing time primarily.

The average raft proposal time (~65ms) is consistent with the time that it takes for the dbInit http request to complete. 65ms x 34 proposals ~= 2.210 secs.

So AFAICT everything is consistent and as-advertised so-to-speak. What surprises me is that the same test with the Hashicorp implementation has easily an order-of-magnitude lower latency.

![image](https://user-images.githubusercontent.com/3630156/91754446-412b8800-eb7e-11ea-8280-eba9462fbbfc.png)

So I am currently moving forward with confidence that I am measuring Dragonboat performance correctly, but that I might not be measuring Hashicorp performance correctly. .5-1ms proposal time seems somewhat absurd... though i did work on a project in 2010 that featured a Java backend that could route messages across a cluster in 5ms.

I am working toward validating the Hashicorp benchmarks in the AWS cloud where there is no chance that HC is taking advantage of some high-speed local networking option. And I am validating the configuration to ensure that there are no corners cut during RAFT WAL persistence / at the DB level.

Re: your code change and concurrency. That sounds very promising. While overall throughput of the backend is an important concern... at the end of the day... each customer (of a consumer application) accesses the service from one client. Latency and uptime (Availability) and Consistency are they only things they care about. Partition-ability is a business concern.

It sounds like a worthy improvement, certainly to include it as a configurable option. I wonder if it's possible to get your latency down to 1ms.

Sincerely,
Jeremy

**Comment by jkassis** [comment]
things look better for dragonboat when run on kubernetes in the cloud... maybe even faster than hashicorp.

![image](https://user-images.githubusercontent.com/3630156/91779038-3fc38500-eba9-11ea-9387-064ef1a9ffef.png)
![image](https://user-images.githubusercontent.com/3630156/91779050-4520cf80-eba9-11ea-8ade-bb40419d610e.png)
![image](https://user-images.githubusercontent.com/3630156/91779062-4a7e1a00-eba9-11ea-9318-5f16e658fe75.png)
![image](https://user-images.githubusercontent.com/3630156/91779073-510c9180-eba9-11ea-8048-ae4c66b65ec3.png)

There are some off-by-one issues in some of these metrics that I haven't quite figured out, but P99 latency < 8ms!

**Comment by lni** [comment]
@jkassis 

checkdisk results provided above was obtained using 1 raft group and 1 client. see the command line I provided in my previous reply. you still need to check why proposal delay jumped from 34ms (in checkdisk) to 65ms in your tests. That >30ms jump (2x) is huge, it didn't just appear for no reason. 

your hashicorp results are definitely wrong as well. you first reported that it took hashicorp 0.6s to complete 34 proposals when your service code is there. you later confirmed that replacing your service with NOOP alone dropped dragonboat's latency from 4.58 seconds on average (3-5 seconds on your screenshot) to 2.210 seconds. so the service code which has nothing to do with dragonboat/hashicorp takes more than 2 seconds to complete. if your service alone is costing more than 2 seconds, how hashicorp can complete everything including your service within 0.6 second?   

for proposal delay itself, in theory the minimum delay one can achieve is one fsync() delay plus one RTT time between nodes. for the vast majority alternative implementations I've checked, in reality for all open source libraries I checked, getting closer to such limit does _NOT_ mean the implementation is better. It means one thing and one thing only - the implementation doesn't have full support for batching. **it is a huge disadvantage.** 

also note that the chart you cited has nothing consistent with what you are doing, the chart is on avg/99.99 percentile latencies when using thousands of clients.

**Comment by lni** [comment]
The issue of LeaderUpdated() repeatedly called with the same LeaderInfo has been fixed in master.

**Comment by jkassis** [comment]
@lni thanks for closing this. and thanks for your most recent post regarding the performance evaluation. I have run out of time to evaluate this as my primary focus, so I am occasionally investigating further. If you like I can keep you up to date with my findings on this ticket or another thread of your choice. until i hear otherwise... i'll post here...

i did an AB test with the same app running on my darwin laptop and running inside an alpine virtual machine on my darwin laptop. the app running in the VM does not show the same performance problem, so this seems to rule out hardware / disk issues.

**Comment by lni** [comment]
@jkassis 

Please see the issues I mentioned in my 2nd Sept reply.
